 # load_digits : 손글씨를 분류
 

## 필요 모듈 임포트 

In [339]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

## 데이터 준비
load_digits 메서드를 사용합니다.



In [340]:
digits = load_digits()
print((dir(digits))) 
print(type(digits))

digits.keys()



['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']
<class 'sklearn.utils.Bunch'>


dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

digits에는 data, target, frame, feature_names, target_names, images,DescR의 정보가 있음을 확인할 수 있다. 

## 데이터 이해하기 
* Feature Data 지정하기
* Label Data 지정하기
* Target Names 출력해 보기
* 데이터 Describe 해 보기

In [341]:
digits_data = digits.data
print(digits_data.shape)
print(digits.feature_names,end= "")

(1797, 64)
['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4', 'pixel_0_5', 'pixel_0_6', 'pixel_0_7', 'pixel_1_0', 'pixel_1_1', 'pixel_1_2', 'pixel_1_3', 'pixel_1_4', 'pixel_1_5', 'pixel_1_6', 'pixel_1_7', 'pixel_2_0', 'pixel_2_1', 'pixel_2_2', 'pixel_2_3', 'pixel_2_4', 'pixel_2_5', 'pixel_2_6', 'pixel_2_7', 'pixel_3_0', 'pixel_3_1', 'pixel_3_2', 'pixel_3_3', 'pixel_3_4', 'pixel_3_5', 'pixel_3_6', 'pixel_3_7', 'pixel_4_0', 'pixel_4_1', 'pixel_4_2', 'pixel_4_3', 'pixel_4_4', 'pixel_4_5', 'pixel_4_6', 'pixel_4_7', 'pixel_5_0', 'pixel_5_1', 'pixel_5_2', 'pixel_5_3', 'pixel_5_4', 'pixel_5_5', 'pixel_5_6', 'pixel_5_7', 'pixel_6_0', 'pixel_6_1', 'pixel_6_2', 'pixel_6_3', 'pixel_6_4', 'pixel_6_5', 'pixel_6_6', 'pixel_6_7', 'pixel_7_0', 'pixel_7_1', 'pixel_7_2', 'pixel_7_3', 'pixel_7_4', 'pixel_7_5', 'pixel_7_6', 'pixel_7_7']

data는 1797개의 이미지로, 한 이미지는 8*8개 총 64이미로 되어있는것을 확인할 수 있었다. 

In [342]:
digits_label = digits.target
print(digits_label.shape)
digits_label


(1797,)


array([0, 1, 2, ..., 8, 9, 8])

In [343]:
print(digits.target_names)

[0 1 2 3 4 5 6 7 8 9]


타겟을 digits_label에 저장하였다.         
data의 숫자와 같은  1797개의 데이터가 있음을 확인할 수 있었다.
target의 이름들은 0부터 9까지의 숫자를 나타내고 있었다. 

In [344]:
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 5620
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

DESC에서는 데이터에 대한 설명을 살펴 볼 수 가 있었다. 

In [345]:
import pandas as pd
digits_df = pd.DataFrame(data=digits_data, columns = digits.feature_names)
digits_df["label"] = digits.target

pandda library를 통해 DataFrame 자료형으로 변형하여 조금 더 쉽게 사용했다.

## train, test 데이터 분리 
모델 학습과 테스트용 문제지와 정답지를 준비해 봅시다.
X_train, X_test, y_train, y_test를 생성하는 방법을 참고해 보세요.

In [346]:
from sklearn.model_selection import train_test_split


d_x_train, d_x_test, d_y_train, d_y_test  = train_test_split(digits_data,
                                                             digits_label,
                                                            test_size = 0.2,
                                                            random_state = 5)

print('d_x_train 개수:',len(d_x_train),'d_x_test 개수:',len(d_x_test))


d_x_train 개수: 1437 d_x_test 개수: 360


In [347]:
d_x_train.shape, d_y_train.shape



((1437, 64), (1437,))

In [348]:
d_x_test.shape, d_y_test.shape

((360, 64), (360,))

In [349]:
d_y_train, d_y_test

(array([9, 0, 0, ..., 6, 8, 3]),
 array([5, 2, 5, 5, 5, 3, 2, 2, 3, 5, 9, 8, 7, 2, 1, 1, 5, 7, 0, 4, 5, 9,
        0, 2, 1, 3, 4, 7, 5, 2, 1, 1, 2, 9, 8, 1, 4, 5, 4, 9, 5, 5, 6, 0,
        1, 7, 2, 9, 7, 7, 3, 9, 5, 1, 6, 6, 7, 7, 8, 1, 6, 1, 3, 6, 1, 3,
        2, 6, 8, 1, 4, 7, 1, 6, 0, 0, 5, 1, 3, 5, 1, 6, 4, 0, 4, 7, 5, 7,
        8, 3, 7, 8, 5, 1, 1, 7, 5, 9, 7, 9, 3, 0, 7, 8, 7, 4, 8, 3, 2, 8,
        5, 2, 7, 4, 4, 8, 9, 7, 4, 5, 0, 5, 9, 8, 2, 3, 2, 4, 4, 8, 0, 5,
        2, 9, 4, 8, 6, 5, 9, 7, 8, 0, 9, 4, 3, 8, 7, 5, 5, 3, 3, 5, 1, 0,
        8, 7, 2, 8, 4, 1, 0, 0, 3, 6, 4, 7, 7, 0, 4, 9, 2, 8, 7, 9, 7, 2,
        0, 3, 3, 8, 5, 7, 5, 6, 8, 4, 1, 5, 1, 1, 6, 9, 9, 9, 8, 6, 4, 6,
        0, 1, 6, 5, 3, 9, 0, 2, 7, 8, 8, 7, 3, 8, 3, 9, 3, 0, 9, 6, 0, 4,
        0, 3, 5, 0, 4, 3, 5, 8, 8, 9, 2, 5, 0, 8, 3, 7, 4, 3, 7, 9, 2, 6,
        1, 2, 1, 7, 0, 7, 5, 0, 6, 4, 1, 8, 3, 0, 8, 9, 2, 2, 5, 2, 6, 6,
        3, 4, 0, 7, 1, 5, 3, 8, 7, 3, 4, 2, 5, 1, 3, 0, 0, 9, 3, 8, 8, 3,
     

무작위로 잘 섞여있는 것을 확인 할 수 있다. 

## 다양한 모델로 학습

1. Decision Tree
2. Random Forest
3. SVM 
4. SGD Classfifier
5. Logistic Regression


### Decsion Tree

In [357]:
from sklearn.tree import DecisionTreeClassifier

d_decision_tree = DecisionTreeClassifier(random_state=32)
print(d_decision_tree._estimator_type)

classifier


In [358]:
d_decision_tree.fit(d_x_train, d_y_train)

DecisionTreeClassifier(random_state=32)

In [359]:
d_d_y_pred = d_decision_tree.predict(d_x_test)
d_d_y_pred

array([5, 2, 8, 5, 5, 9, 2, 8, 3, 5, 9, 8, 7, 2, 1, 8, 5, 3, 0, 4, 5, 9,
       8, 2, 1, 3, 4, 7, 5, 2, 7, 1, 2, 9, 8, 1, 4, 9, 7, 9, 5, 5, 6, 0,
       1, 7, 2, 9, 7, 7, 7, 9, 5, 1, 1, 6, 7, 7, 8, 1, 6, 1, 3, 6, 1, 3,
       2, 6, 8, 1, 4, 7, 1, 6, 0, 0, 5, 1, 3, 8, 1, 6, 4, 0, 4, 7, 7, 7,
       8, 3, 7, 8, 5, 1, 1, 7, 5, 9, 4, 9, 9, 0, 7, 8, 7, 5, 8, 3, 2, 8,
       5, 1, 7, 4, 9, 8, 9, 7, 4, 8, 0, 5, 9, 8, 2, 3, 2, 4, 7, 8, 0, 5,
       2, 9, 4, 9, 8, 5, 9, 2, 1, 0, 9, 4, 3, 8, 7, 5, 9, 3, 3, 5, 1, 0,
       8, 7, 5, 8, 4, 1, 0, 0, 3, 6, 9, 3, 7, 0, 5, 9, 2, 8, 7, 9, 7, 2,
       0, 3, 3, 8, 5, 7, 5, 6, 8, 9, 1, 5, 3, 1, 6, 9, 9, 9, 8, 6, 4, 6,
       0, 1, 6, 5, 3, 9, 0, 2, 7, 8, 1, 7, 3, 1, 3, 9, 3, 0, 9, 6, 0, 9,
       0, 2, 5, 0, 4, 8, 5, 5, 8, 9, 2, 5, 0, 8, 3, 7, 4, 3, 3, 9, 2, 6,
       1, 7, 1, 7, 0, 7, 5, 0, 6, 4, 8, 8, 3, 0, 8, 9, 6, 2, 5, 2, 6, 4,
       3, 4, 0, 7, 1, 5, 4, 2, 4, 3, 4, 7, 5, 1, 3, 0, 0, 9, 3, 8, 8, 3,
       9, 8, 6, 6, 2, 6, 7, 5, 3, 1, 9, 7, 5, 4, 5,

In [360]:
from sklearn.metrics import accuracy_score

d_d_accuracy = accuracy_score(d_y_test, d_d_y_pred)
d_d_accuracy

0.8527777777777777

In [361]:
print(classification_report(d_y_test, d_d_y_pred))

              precision    recall  f1-score   support

           0       1.00      0.91      0.96        35
           1       0.82      0.89      0.85        35
           2       0.88      0.81      0.84        36
           3       0.89      0.84      0.86        37
           4       0.85      0.72      0.78        32
           5       0.88      0.83      0.85        46
           6       0.96      0.90      0.93        30
           7       0.83      0.83      0.83        42
           8       0.78      0.84      0.81        38
           9       0.71      1.00      0.83        29

    accuracy                           0.85       360
   macro avg       0.86      0.86      0.85       360
weighted avg       0.86      0.85      0.85       360



### Random Forest

In [362]:
from sklearn.ensemble import RandomForestClassifier


d_random_forest = RandomForestClassifier(random_state=32)
d_random_forest.fit(d_x_train, d_y_train)
d_r_y_pred = d_random_forest.predict(d_x_test)

print(classification_report(d_y_test, d_r_y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.99        35
           1       0.97      1.00      0.99        35
           2       0.97      1.00      0.99        36
           3       1.00      0.95      0.97        37
           4       0.97      0.97      0.97        32
           5       0.94      0.96      0.95        46
           6       1.00      0.97      0.98        30
           7       0.98      0.98      0.98        42
           8       0.92      0.95      0.94        38
           9       0.93      0.93      0.93        29

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



### SVM  

In [363]:
from sklearn import svm
d_svm_model = svm.SVC()

print(d_svm_model._estimator_type)

classifier


In [364]:
d_svm_model.fit(d_x_train, d_y_train)
d_svm_y_pred = d_svm_model.predict(d_x_test)
print(classification_report(d_y_test,d_svm_y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        35
           1       0.97      1.00      0.99        35
           2       1.00      1.00      1.00        36
           3       1.00      0.95      0.97        37
           4       1.00      1.00      1.00        32
           5       0.96      0.98      0.97        46
           6       1.00      1.00      1.00        30
           7       1.00      0.98      0.99        42
           8       0.95      0.95      0.95        38
           9       0.90      0.93      0.92        29

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



### SGD Classfifier

In [365]:
from sklearn.linear_model import SGDClassifier
d_sgd_model = SGDClassifier()

print(d_sgd_model._estimator_type)

classifier


In [366]:
d_sgd_model.fit(d_x_train, d_y_train)
d_sgd_y_pred = d_sgd_model.predict(d_x_test)
print(classification_report(d_y_test,d_sgd_y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.99        35
           1       0.78      1.00      0.88        35
           2       1.00      0.97      0.99        36
           3       1.00      0.95      0.97        37
           4       0.97      0.91      0.94        32
           5       1.00      0.91      0.95        46
           6       1.00      0.97      0.98        30
           7       0.98      0.98      0.98        42
           8       0.90      0.92      0.91        38
           9       0.90      0.90      0.90        29

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



### Logistic Regression

In [413]:
from sklearn.linear_model import LogisticRegression
d_logistic_model = LogisticRegression(max_iter = 10000)

print(d_logistic_model._estimator_type)

classifier


In [414]:
d_logistic_model.fit(d_x_train, d_y_train)
d_l_y_pred = d_logistic_model.predict(d_x_test)
print(classification_report(d_y_test,d_l_y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        35
           1       0.89      0.97      0.93        35
           2       0.95      1.00      0.97        36
           3       0.97      0.95      0.96        37
           4       0.97      0.91      0.94        32
           5       0.98      0.93      0.96        46
           6       1.00      0.97      0.98        30
           7       0.98      0.98      0.98        42
           8       0.97      0.92      0.95        38
           9       0.88      0.97      0.92        29

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.96      0.96      0.96       360



## 모델 평가 
학습된 모델들의 테스트데이터 예측 결과를 어떻게 해석해야 할까요?        
모델의 성능을 평가하는 지표로는 무엇이 좋을까요?            
sklearn.metrics 에서 제공하는 평가지표 중 적절한 것을 선택해 보세요.           
선택하신 이유도 설명해 주세요.

In [369]:
from sklearn.metrics import confusion_matrix
confusion_matrix(d_y_test, d_l_y_pred)

array([[35,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 34,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 0,  1, 35,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  1, 35,  0,  1,  0,  0,  0,  0],
       [ 0,  2,  0,  0, 29,  0,  0,  1,  0,  0],
       [ 0,  0,  0,  0,  1, 43,  0,  0,  0,  2],
       [ 0,  0,  0,  0,  0,  0, 29,  0,  1,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 41,  0,  1],
       [ 0,  1,  1,  0,  0,  0,  0,  0, 35,  1],
       [ 0,  1,  0,  0,  0,  0,  0,  0,  0, 28]])

숫자 데이터는 precsion과 recall 둘다 중요하다고 생각합니다.
기본적으로 숫자는 빈도가 특별하게 높은 숫자가 존재하지않으며
특정한 숫자를 분류하는것이 중요한 목적이 아니라 전체적인 판단능력이 좋아야한고 생각합니다.

digits 데이터에서는 precison과 recall이 양측다 높고 자연스럽게 f-socre도 높은
SVM을 선택하는것이 적절하다고 생각합니다.

 # load_digits : load_wine : 와인을 분류
 

## 필요 모듈 임포트 

In [370]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

## 데이터 준비
load_wine 메서드를 사용합니다.



In [371]:
wine = load_wine()
print((dir(wine))) 
print(type(wine))

wine.keys()



['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']
<class 'sklearn.utils.Bunch'>


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

wine에는 DESCR,data, feature_names, frame, target, target_naems      
의 정보가 있음을 확인했다. 

## 데이터 이해하기 
* Feature Data 지정하기
* Label Data 지정하기
* Target Names 출력해 보기
* 데이터 Describe 해 보기

In [ ]:
wine_data= wine.data
print(wine_data.shape)
print(wine.feature_names,end ="")


data는 178개로 한 와인 데이터는 13가지  feature를 가지는것을 확인할 수 있었다.    
종류에는 alcohol, ash 등이 있었다. 

In [ ]:
wine_label = wine.target
print(wine_label.shape)
wine_label

In [ ]:
print(wine.target_names)

타겟을 wine_label에 저장하였따.
wine의 데이터숫자와 같은 178개의 데이터가 있음을 확인할수있었다.
target의 이름은 class 0,1,2으로 3종류임을 확인했다.

In [373]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

데이터는 이탈리아의 세 다른 지역의 농부들이 지은 와인에 대한 화학적 분석 결과물이라고 한다. 

In [374]:
import pandas as pd 
wine_df = pd.DataFrame(data = wine_data, columns = wine.feature_names)
wine_df["label"]  = wine.target

pandda library를 통해 DataFrame 자료형으로 변형하여 조금 더 쉽게 사용할수도 있다. 

## train, test 데이터 분리 

In [375]:
from sklearn.model_selection import train_test_split


w_x_train, w_x_test, w_y_train, w_y_test  = train_test_split(wine_data,
                                                             wine_label,
                                                            test_size = 0.2,
                                                            random_state = 37)

print('w_x_train 개수:',len(w_x_train),'w_x_test 개수:',len(w_x_test))


w_x_train 개수: 142 w_x_test 개수: 36


In [376]:
w_y_train, w_y_test


(array([1, 2, 0, 0, 2, 1, 0, 1, 2, 1, 2, 1, 2, 0, 2, 1, 0, 2, 1, 2, 2, 1,
        0, 1, 2, 2, 0, 1, 1, 1, 1, 2, 1, 2, 0, 0, 2, 0, 1, 2, 1, 0, 0, 1,
        2, 1, 2, 2, 0, 0, 0, 0, 1, 1, 0, 1, 2, 1, 0, 1, 0, 0, 1, 1, 1, 1,
        1, 1, 1, 2, 0, 1, 1, 0, 1, 2, 0, 2, 0, 1, 0, 1, 2, 1, 1, 0, 0, 1,
        1, 2, 0, 0, 1, 2, 1, 0, 1, 1, 2, 0, 0, 1, 1, 0, 0, 0, 2, 0, 2, 1,
        2, 0, 0, 1, 0, 1, 2, 0, 2, 1, 2, 0, 0, 2, 0, 1, 2, 0, 2, 2, 0, 1,
        1, 1, 1, 0, 2, 1, 0, 1, 1, 2]),
 array([1, 0, 0, 2, 0, 2, 0, 1, 1, 1, 2, 0, 0, 2, 2, 0, 1, 2, 0, 1, 0, 1,
        2, 1, 2, 0, 1, 1, 0, 0, 1, 2, 1, 2, 1, 1]))

무작위로 잘 섞여있는 것을 확인 할 수 있다. 

## 다양한 모델로 학습

1. Decision Tree
2. Random Forest
3. SVM 
4. SGD Classfifier
5. Logistic Regression


### Decsion Tree

In [377]:
from sklearn.tree import DecisionTreeClassifier

w_decision_tree = DecisionTreeClassifier(random_state=7)
print(w_decision_tree._estimator_type)

classifier


In [378]:
w_decision_tree.fit(w_x_train, w_y_train)

DecisionTreeClassifier(random_state=7)

In [379]:
w_d_y_pred = w_decision_tree.predict(w_x_test)
w_d_y_pred

array([1, 0, 0, 2, 1, 2, 0, 1, 1, 1, 1, 0, 0, 2, 2, 0, 1, 2, 0, 1, 0, 1,
       2, 0, 2, 0, 0, 1, 0, 0, 1, 2, 1, 2, 0, 2])

In [380]:
from sklearn.metrics import accuracy_score
w_d_accuracy = accuracy_score(w_y_test, w_d_y_pred)
w_d_accuracy

0.8333333333333334

In [381]:
print(classification_report(w_y_test,w_d_y_pred))

              precision    recall  f1-score   support

           0       0.79      0.92      0.85        12
           1       0.83      0.71      0.77        14
           2       0.90      0.90      0.90        10

    accuracy                           0.83        36
   macro avg       0.84      0.84      0.84        36
weighted avg       0.84      0.83      0.83        36



### Random Forest

In [382]:
from sklearn.ensemble import RandomForestClassifier
w_random_forest = RandomForestClassifier(random_state = 37)
w_random_forest.fit(w_x_train, w_y_train)
w_r_y_pred = w_random_forest.predict(w_x_test)
print(classification_report(w_y_test,w_r_y_pred))

              precision    recall  f1-score   support

           0       0.92      1.00      0.96        12
           1       1.00      0.93      0.96        14
           2       1.00      1.00      1.00        10

    accuracy                           0.97        36
   macro avg       0.97      0.98      0.97        36
weighted avg       0.97      0.97      0.97        36



### SVM  

In [383]:
from sklearn import svm
w_svm_model = svm.SVC()
print(w_svm_model._estimator_type)

classifier


In [384]:
w_svm_model.fit(w_x_train,w_y_train)
w_svm_y_pred = w_svm_model.predict(w_x_test)
print(classification_report(w_y_test, w_svm_y_pred))

              precision    recall  f1-score   support

           0       0.64      0.75      0.69        12
           1       0.62      0.57      0.59        14
           2       0.33      0.30      0.32        10

    accuracy                           0.56        36
   macro avg       0.53      0.54      0.53        36
weighted avg       0.55      0.56      0.55        36



### SGD Classfifier

In [385]:
from sklearn.linear_model import SGDClassifier
w_sgd_model = SGDClassifier()
print(w_sgd_model._estimator_type)  # 분류모델 확인 

classifier


In [386]:
w_sgd_model.fit(w_x_train, w_y_train)
w_sgd_y_pred = w_sgd_model.predict(w_x_test)
print(classification_report(w_y_test,w_sgd_y_pred))

              precision    recall  f1-score   support

           0       0.82      0.75      0.78        12
           1       1.00      0.07      0.13        14
           2       0.38      0.90      0.53        10

    accuracy                           0.53        36
   macro avg       0.73      0.57      0.48        36
weighted avg       0.77      0.53      0.46        36



In [387]:
w_sgd_model.fit(w_x_train, w_y_train)
w_sgd_y_pred = w_sgd_model.predict(w_x_train)

In [388]:
w_sgd_y_pred

array([1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1,
       0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1,
       1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1,
       1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 1, 0, 1, 1, 1])

sgd classififer에서 class 2를 예측한 값이 하나도 존재하지 않게되어 분류모델로서는 최악으로 적합되었다. 

In [389]:
w_y_test

array([1, 0, 0, 2, 0, 2, 0, 1, 1, 1, 2, 0, 0, 2, 2, 0, 1, 2, 0, 1, 0, 1,
       2, 1, 2, 0, 1, 1, 0, 0, 1, 2, 1, 2, 1, 1])

### Logistic Regression

In [390]:
from sklearn.linear_model import LogisticRegression
w_logistic_model = LogisticRegression(max_iter = 10000)
print(w_logistic_model)

LogisticRegression(max_iter=10000)


In [391]:
w_logistic_model.fit(w_x_train, w_y_train)
w_l_y_pred = w_logistic_model.predict(w_x_test)
print(classification_report(w_y_test,w_l_y_pred))

              precision    recall  f1-score   support

           0       0.80      1.00      0.89        12
           1       0.92      0.79      0.85        14
           2       1.00      0.90      0.95        10

    accuracy                           0.89        36
   macro avg       0.91      0.90      0.89        36
weighted avg       0.90      0.89      0.89        36



## 모델 평가 
학습된 모델들의 테스트데이터 예측 결과를 어떻게 해석해야 할까요?        
모델의 성능을 평가하는 지표로는 무엇이 좋을까요?            
sklearn.metrics 에서 제공하는 평가지표 중 적절한 것을 선택해 보세요.           
선택하신 이유도 설명해 주세요.

sgd는 test데이터에 대해서 class2로 분류한 값이 전혀없는만큼 최악의 성능을 보였습니다. 
와인도 위의 digits와 비슷한 상황으로 전체적인 성능이 중요하다고 생각합니다.
precision, recall 양쪽에서 우수한 성능(그리고 당연히 f1-score까지)보여준
random forest가 우수한 모델이라고 생각하빈다. 

 # load_breast_cancer : 유방암 여부를 진단
 

## 필요 모듈 임포트 

In [392]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

## 데이터 준비
load_breast_cancer 메서드를 사용합니다.



In [393]:
cancer = load_breast_cancer()
print((dir(cancer)))
print(type(cancer))

cancer.keys()

['DESCR', 'data', 'feature_names', 'filename', 'frame', 'target', 'target_names']
<class 'sklearn.utils.Bunch'>


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])

cancer에는 DESCR,data,feature_names,filename, frame, target 등의 정보가 있음을    
확인했다.

## 데이터 이해하기 
* Feature Data 지정하기
* Label Data 지정하기
* Target Names 출력해 보기
* 데이터 Describe 해 보기

In [394]:
cancer_data = cancer.data
print(cancer_data.shape)
print(cancer.feature_names, end = "")

(569, 30)
['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']

데이터는 총 569개로 30가지 feature를 가지는 것을 확인할 수 있었음    
종류에는 mean radius, mean texture, worst symmetry등이 있었음 

In [395]:
cancer_label = cancer.target
print(cancer_label.shape)
cancer_label

(569,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [396]:
print(cancer.target_names)

['malignant' 'benign']


타겟을 cancer_label에 저장하였다. 
cancer의 데이터 숫자와 같은 569개임을 확인하였다.
target의 이름은 malignant, benign 양성종양과 악성종양임을 확인할 수 있었다.

In [397]:
print(cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [398]:
import pandas as pd
cancer_df = pd.DataFrame(data = cancer_data, columns = cancer.feature_names)
cancer_df["label"] = cancer.target

panda를통해 DataFrame 자료형으로 변형도 해보았다. 

## train, test 데이터 분리 

In [422]:
from sklearn.model_selection import train_test_split

c_x_train, c_x_test, c_y_train, c_y_test = train_test_split(cancer_data,
                                                           cancer_label,
                                                           test_size = 0.2,
                                                           random_state = 2)
print('c_x_train 개수:', len(c_x_train), 'c_x_test 개수:',len(c_x_test))

c_x_train 개수: 455 c_x_test 개수: 114


In [400]:
c_y_train, c_y_test

(array([0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0,
        0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1,
        0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1,
        0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0,
        1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1,
        1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0,
        1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0,
        1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1,
        1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1,
        0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1,
        1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1,
        1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 

## 다양한 모델로 학습

1. Decision Tree
2. Random Forest
3. SVM 
4. SGD Classfifier
5. Logistic Regression


### Decsion Tree

In [423]:
from sklearn.tree import DecisionTreeClassifier
c_decision_tree = DecisionTreeClassifier(random_state = 1)
print(c_decision_tree._estimator_type)

classifier


In [424]:
c_decision_tree.fit(c_x_train,c_y_train)

DecisionTreeClassifier(random_state=1)

In [425]:
c_d_y_pred = c_decision_tree.predict(c_x_test)
c_d_y_pred

array([1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0,
       0, 0, 1, 0])

In [426]:
from sklearn.metrics import accuracy_score
c_d_accuracy = accuracy_score(c_y_test,c_d_y_pred)
c_d_accuracy

0.9122807017543859

In [427]:
print(classification_report(c_y_test,c_d_y_pred))

from sklearn.metrics import confusion_matrix
confusion_matrix(c_y_test, c_d_y_pred)

              precision    recall  f1-score   support

           0       0.84      0.96      0.90        45
           1       0.97      0.88      0.92        69

    accuracy                           0.91       114
   macro avg       0.91      0.92      0.91       114
weighted avg       0.92      0.91      0.91       114



array([[43,  2],
       [ 8, 61]])

### Random Forest

In [428]:
from sklearn.ensemble import RandomForestClassifier

c_random_forest = RandomForestClassifier(random_state = 1)
c_random_forest.fit(c_x_train, c_y_train)
c_r_y_pred = c_random_forest.predict(c_x_test)

print(classification_report(c_y_test, c_r_y_pred))

from sklearn.metrics import confusion_matrix
confusion_matrix(c_y_test, c_r_y_pred)

              precision    recall  f1-score   support

           0       0.91      0.96      0.93        45
           1       0.97      0.94      0.96        69

    accuracy                           0.95       114
   macro avg       0.94      0.95      0.95       114
weighted avg       0.95      0.95      0.95       114



array([[43,  2],
       [ 4, 65]])

### SVM  

In [429]:
from sklearn import svm
c_svm_model = svm.SVC()

print(c_svm_model._estimator_type)

classifier


In [430]:
c_svm_model.fit(c_x_train,c_y_train)
c_svm_y_pred = c_svm_model.predict(c_x_test)
print(classification_report(c_y_test,c_svm_y_pred))

from sklearn.metrics import confusion_matrix
confusion_matrix(c_y_test, c_svm_y_pred)

              precision    recall  f1-score   support

           0       0.90      0.84      0.87        45
           1       0.90      0.94      0.92        69

    accuracy                           0.90       114
   macro avg       0.90      0.89      0.90       114
weighted avg       0.90      0.90      0.90       114



array([[38,  7],
       [ 4, 65]])

### SGD Classfifier

In [431]:
from sklearn.linear_model import SGDClassifier
c_sgd_model = SGDClassifier()
print(c_sgd_model._estimator_type)

classifier


In [432]:
c_sgd_model.fit(c_x_train, c_y_train)
c_sgd_y_pred = c_sgd_model.predict(c_x_test)
print(classification_report(c_y_test,c_sgd_y_pred))

from sklearn.metrics import confusion_matrix
confusion_matrix(c_y_test, c_sgd_y_pred)

              precision    recall  f1-score   support

           0       1.00      0.69      0.82        45
           1       0.83      1.00      0.91        69

    accuracy                           0.88       114
   macro avg       0.92      0.84      0.86       114
weighted avg       0.90      0.88      0.87       114



array([[31, 14],
       [ 0, 69]])

### Logistic Regression

In [433]:
from sklearn.linear_model import LogisticRegression
c_logistic_model = LogisticRegression(max_iter = 10000)

print(c_logistic_model._estimator_type)

classifier


In [434]:
c_logistic_model.fit(c_x_train,c_y_train)
c_l_y_pred = c_logistic_model.predict(c_x_test)
print(classification_report(c_y_test,c_l_y_pred))

from sklearn.metrics import confusion_matrix
confusion_matrix(c_y_test, c_l_y_pred)

              precision    recall  f1-score   support

           0       0.91      0.91      0.91        45
           1       0.94      0.94      0.94        69

    accuracy                           0.93       114
   macro avg       0.93      0.93      0.93       114
weighted avg       0.93      0.93      0.93       114



array([[41,  4],
       [ 4, 65]])

## 모델 평가 
학습된 모델들의 테스트데이터 예측 결과를 어떻게 해석해야 할까요?        
모델의 성능을 평가하는 지표로는 무엇이 좋을까요?            
sklearn.metrics 에서 제공하는 평가지표 중 적절한 것을 선택해 보세요.           
선택하신 이유도 설명해 주세요.

유방암 데이터는 지금까지와의 데이터와 다른 상황입니다.
암 검진을 받는 환자는 실제 환자가 아니더라도 악성종양이라고 판단받는것이 중요합니다.
왜냐하면 그렇지 않으면 암 환자인데도 검진에서 건강한 사람이라고 판정되면
후속 치료나 진단을 받지못하게되기 때문입니다. 
즉 지금 상황에서는 recall이 높은것이 더 중요합니다. 

특히 0(maligannt)에 대한 recall이 중요합니다. 
그러므로 그 것이 높은 decision tree와 random forest가 좋은 모델입니다.

나머지 precision과 recall(그리고 f1-socre에서도) random forest가     
더 좋은 성능을 보였기 떄문에 random forest가 최고의 모델이라고 고를 수 있을것같습니다.

## 회고 

원래 f1_score는 recall과 precsion의 조화평균으로 양측을 모두 잡으려는 것을 하나의 기준으로 세운것으로 알고있습니다. 그런데 f1-score는 조화평균을 계산할때의 비중이 1 대 1입니다. 이것을 좀더 일반화하여 본인이 recall과 precsion에 한 측에 대해 더 가중치를 부여하는 방법을 generalized f1-score라고 부릅니다. 그러나 이번 기회에서는 모든 모델의 성능이 모든면에서 우월한 경우가 많아서 활용할 기회가 없어서 아쉬웠습니다. 